src/platform/platform_unix.cpp

```diff
@@ -175,11 +177,46 @@ sigset_t mp::platform::make_and_block_signals(const std::vector<int>& sigs)
     return sigset;
 }
 
-std::function<int()> mp::platform::make_quit_watchdog()
+template <class Rep, class Period>
+timespec make_timespec(std::chrono::duration<Rep, Period> duration)
 {
-    return [sigset = make_and_block_signals({SIGQUIT, SIGTERM, SIGHUP})]() {
-        int sig = -1;
-        sigwait(&sigset, &sig);
-        return sig;
+    const auto seconds = std::chrono::duration_cast<std::chrono::seconds>(duration);
+    return timespec{seconds.count(), std::chrono::duration_cast<std::chrono::nanoseconds>(duration - seconds).count()};
+}
+
+std::function<std::optional<int>(const std::function<bool()>&)> mp::platform::make_quit_watchdog(
+    const std::chrono::milliseconds& period)
+{
+    return [sigset = make_and_block_signals({SIGQUIT, SIGTERM, SIGHUP, SIGUSR2}),
+            period](const std::function<bool()>& condition) -> std::optional<int> {
+        std::mutex sig_mtx;
+        std::condition_variable sig_cv;
+        int sig = SIGUSR2;
+
+        // A signal generator that triggers after `timeout`
+        AutoJoinThread signaler([&sig_mtx, &sig_cv, &sig, &period, signalee = pthread_self()] {
+            std::unique_lock lock(sig_mtx);
+            while (!sig_cv.wait_for(lock, period, [&sig] { return sig != SIGUSR2; }))
+            {
+                pthread_kill(signalee, SIGUSR2);
+            }
+        });
+
+        // wait on signals and condition
+        int ret = SIGUSR2;
+        while (ret == SIGUSR2 && condition())
+        {
+            // can't use sigtimedwait since macOS doesn't support it
+            sigwait(&sigset, &ret);
+        }
+
+        {
+            std::lock_guard lock(sig_mtx);
+            sig = ret == SIGUSR2 ? 0 : ret;
+        }
+        sig_cv.notify_all();
+
+        // if sig is SIGUSR2 then we know we're exiting because condition() was false
+        return ret == SIGUSR2 ? std::nullopt : std::make_optional(sig);
```

Comment:
And then we could use a single variable here to simplify this slightly:

```suggestion
        return sig ? std::make_optional(sig) : std::nullopt;
```